File: src/main/java/no/nav/pensjon/regler/domain/grunnlag/Uttaksgrad.java
package no.nav.pensjon.regler.domain.grunnlag;

import no.nav.pensjon.regler.domain.util.DateCompareUtil;

import java.io.Serializable;
import java.util.Date;

public class Uttaksgrad implements Comparable<Uttaksgrad>, Serializable {

    private static final long serialVersionUID = -8963723653801267710L;
    private Date fomDato;
    private Date tomDato;
    private int uttaksgrad;

    public Uttaksgrad() {
    }

    public Uttaksgrad(Uttaksgrad uttaksgrad) {
        fomDato = uttaksgrad.fomDato;
        tomDato = uttaksgrad.tomDato;
        this.uttaksgrad = uttaksgrad.uttaksgrad;
    }

    /**
     * Sammenligner mht fomDato - nyeste fomDato vinner
     */
    @Override
    public int compareTo(Uttaksgrad grad) {
        return DateCompareUtil.compareTo(getFomDato(), grad.getFomDato());
    }

    public Date getFomDato() {
        return fomDato;
    }

    public void setFomDato(Date fomDato) {
        this.fomDato = fomDato;
    }

    public Date getTomDato() {
        return tomDato;
    }

    public void setTomDato(Date tomDato) {
        this.tomDato = tomDato;
    }

    public int getUttaksgrad() {
        return uttaksgrad;
    }

    public void setUttaksgrad(int uttaksgrad) {
        this.uttaksgrad = uttaksgrad;
    }
}

File: src/main/kotlin/no/nav/pensjon/regler/domain/grunnlag/Uttaksgrad.kt
package no.nav.pensjon.regler.domain.grunnlag

import java.io.Serializable
import java.util.*
import java.time.LocalDate

// TODO Skriv om compare-funksjonen
class Uttaksgrad : Serializable, Comparable<Uttaksgrad> {
    @Deprecated("Use fomDatoLd instead")
    var fomDato: Date? = null
    var fomDatoLd: LocalDate? = null
    @Deprecated("Use tomDatoLd instead")
    var tomDato: Date? = null
    var tomDatoLd: LocalDate? = null
    var uttaksgrad = 0

    override fun compareTo(other: Uttaksgrad): Int {
        return DateCompareUtil.compareTo(fomDato, other.fomDato)
    }

}

object DateCompareUtil {
    const val BEFORE = -1
    const val EQUAL = 0
    const val AFTER = 1

    /**
     * Returnerer resultat i forhold til compareTo standard
     *
     * @param first - dato fra kallende klasse
     * @param second - dato fra argument til compareTo
     * @return
     */
    fun compareTo(first: Date?, second: Date?): Int {
        // null sorteres foran
        if (first == null) {
            return if (second == null) {
                EQUAL
            } else {
                BEFORE
            }
        } else if (second == null) {
            return AFTER
        }
        return first.compareTo(second)
    }
}